### ISSUE TITLE
Extract citation parser into its own Python package

### ISSUE THREAD

**Issue by mattdahl** [freelawproject/courtlistener#1521]
It would be nice if our citation parser were decoupled from the rest of the CourtListener Django infrastructure. Doing so would both (1) make it easier for third-parties to use and integrate our tools into their own apps and (2) incentivize such parties to contribute improvements to the parser (when they otherwise might not, at least so long as making changes to it requires working within CourtListener).

@slbayer mentioned in #1338 that he was doing something like this -- if you are able to share any of your work here, please feel free to do so.

What I think should be extracted into its own package is the following:
- Everything in `find_citations.py`
- Most everything except the `as_html()` methods in `models.py`
- The appropriate tests in `tests.py` (i.e., those related to parsing, but not matching or anything else)

Here are some of the challenges I envision at this point:
- Maintaining access to each court's citation strings, which are currently stored in the Django DB. (https://github.com/freelawproject/courtlistener/blob/master/cl/citations/find_citations.py#L112.) Perhaps we could put these in the `reporters_db` instead? Would that make sense? Or put them directly in this (to be named) citation package? These are just static strings, right? (CL would then import them itself as it saw fit.)
- Removing the `as_html()` methods from each `Citation` class and then grafting them on after import. I see no reason why anyone else would want such a method (as the contents of each method are bespoke to CL), but seem a little ugly to me to have to then stick them on later once we import the classes into CL. We would also have to do this for a couple of other methods, like our `to_model()` method, which translates `Citation` objects into Django `search.Citation` models. Alternatively, we could adopt a more functional approach on the CL-side, i.e., not try to put all these methods back onto each object post-hoc but turn them into freestanding helper functions instead. (Such a departure from object-oriented design would cause me pain, though.)
- Designing an appropriate API. The easiest thing to do would be to just expose the same `get_citations()` method that we currently have, and tell people, "it works like so: just give it some text, and get structured citation entities back, conditional on these options that we already have." For an initial version, that seems fine to me, but should we think about designing its usage any differently down the road?

Please let me know if anything I've written is unclear, and chime in with other things to consider.

**Comment by mlissner** [comment]
@jcushman, would something like this be useful to you? I imagine you guys just made something really similar that you're using, right?

**Comment by mlissner** [comment]
@fbennett, I suppose you should be looped in too, in case something like this is useful to your universe?

**Comment by mlissner** [comment]
A couple quick, more substantive replies:

> Perhaps we could put these in the reporters_db instead? 

Seems like a fine solution to me, yes, or putting them into the to-be-named citation solution could work too. As you can see from the code you linked to, getting them from the DB is pretty annoying. They don't change.

> ....`as_html() methods...

Yeah, I don't see a particularly elegant solution here code-wise, but I agree that they probably don't belong in the new system, since they contain CL-specific code. I think I trust your judgment on this one, Matt.

> Designing an appropriate API.

From a change perspective, it'd be nice to have the most minimal thing that yanks out the current citation code and swaps in the new system. That probably means that we try not to touch the API of get_citations if we can, at least not in the first version. Maybe one PR to swap in the new system, and another use it more elegantly.

A couple other stray thoughts:

1. Performance is always something to think about when doing citations. I think most of our citation-related performance problems are on the `match_citations` side, but it's something to keep in mind as we layer on more and more regexes and abstraction.

1. We're doing a lot with docker these days. One way to approach the architecture is to have a python package that CL depends on — that could be fine, and it's how we handle things like Juriscraper. Another way to handle this would be to wrap citation extraction in a microservice. We're migrating to this for OCR and document extraction, but I worry about how it'd perform if we did it here. Still, it'd be nice to have an HTTP endpoint on this from an integration and encapsulation perspective, maybe.

1. Related to the above, we've thought over the years about service for finding citations in various things. Like, could we make an API for it, say? Or a little webpage, where you could post text or a PDF? Not sure if any of this impacts this project, but it's a stray thought that rattles around from time to time. CARA, from CaseText, is a really sophisticated example of this, but funnily enough, the simple version of this — "Might these citations have typos?" — has been on our to do list for a decade or so.

A todo (I'll add these as checkboxes so we remember to get around to them before closing the issue later):

 - [ ] [Issues related to citations](https://github.com/freelawproject/courtlistener/issues?q=is%3Aopen+is%3Aissue+label%3Acitator) might be good to review, and eventually will need to be swapped to the new repo.

**Comment by slbayer** [comment]
As @mattdahl notes, I've done this independently and I can testify that it's really, really useful. 

I can try to share the modifications I made to make this happen, but due to other issues with the guts of the citation finder that I also reported, my version of the code has diverged from yours, and I'm not sure there would be any point. What I **can** do is describe, in detail, what had to change. If you'd like me to explore sharing the modifications with you, I'll talk to my company lawyers. My changes were forked off of your archive revision eb2c6551e049f14bd111fe5547de470d35ebd76a.

1. For code, I kept find_citations.py, match_citations.py, cl/citations/reporter_tokenizer.py, cl/lib/roman.py, cl/citations/models.py, and a small portion of cl/search/models.py.  From cl/search/fixtures, I kept court_data_new3.json, court_data_new2.json, psc_court.json, court_data_new_4.json, court_data_new.json, court_data.json. I loaded these via a stub I added in cl/search/models.py.
2. In find_citations.py, I replaced `django.utils.timezone.now` with `datetime.datetime.now`.
3. I did, indeed, remove the HTML generation, which allowed me to remove all references to `juriscraper`
4. Significantly, because I was using the citation finder for character-offset-based NLP applications, I needed to add a whole bunch of stuff to find_citations.py and reporter_tokenizer.py to track the character offsets of each token. I also chose to convert the citation finder, internally, to a class in order to encapsulate the parser state so I didn't need to pass around all the function arguments I would have needed.
5. In match_citations.py, I stubbed out a chunk of the code that required the CL database in order to do the citation matching (e.g., `best_case_name()`, which now simply produces the citation's `base_citation()`). Removed references to `django.conf.settings`, `cl.lib.sunburnt`, `cl.custom_filters.templatetags.text_filters`. Also, did not try to return the opinion from the database, but rather created a canonical one and tracked it locally.
6. In cl/citation/models.py, I removed `to_model()`, and added local support to store the word and character offsets in the Opinion object. Also removed references to `cl.citations.utils`, `cl.search.models.Citation`.
7. cl/search/models.py was almost completely eviscerated; I just needed enough of it to mimic the Opinion object the list of all Court objects. It was here that I added a stub to load all the fixtures for the courts.
8. The only dependency I required was reporters-db.

Finally, I migrated the whole sub-package to Python 3, which I recommend highly if you want it to be useful separately. There isn't a lot of work there.

In the grand scheme of things, this wasn't difficult. The challenge you'll have is figuring out how to set it up so that the more powerful, Django-based behavior is available when the citation finder is used within CourtListener.

**Comment by jcushman** [comment]
> @jcushman, would something like this be useful to you? I imagine you guys just made something really similar that you're using, right?

Yes, this does sound useful! We're using a less-smart thing at the moment that just counts instances of `<volume> <reporter string> <page>` and doesn't try to extract all that other great citation metadata we could be getting. I'd rather benefit from and contribute to this engine, which is already closer to where we want to be.

There could be issues with adapting a standalone package to different needs (we'd have to tweak things to deal with OCR-specific issues like periods showing up as commas for example), but I bet the community still gains a lot more from monkeypatching features into a common standalone package and upstreaming what we can than from copying and pasting code around.

BTW I'm working on an extractor at the moment that uses hyperscan instead of Python's regex engine for the initial pass of extracting candidate citation strings. Hyperscan is an x86-only library developed at Intel for running regex watchlists on routers, so it's very fast at running lots of regexes in parallel. Like, I found it faster at running hundreds of regexes on a case than python was at running one simple regex, let along a very slow regex like `(A|B|C...)`. This means it becomes feasible to tune separate regexes for each reporter like `\d+-NMSC-\d+`, `\d{4} WL \d{6,}`, `Weirdo (?P<volume>\d+) (?P<page>\d+)`.

I don't have a feel yet for how useful that might be -- it would probably be a nice optimization of your tokenizer function for starters, but does it change anything more broadly in the design? Maybe not! It's neat tech anyway. :)

**Comment by mlissner** [comment]
Wow, thanks for the great replies everybody. This is all really good context to have as Matt starts pulling this out. 

Two quick replies:

1. @slbayer we're migrated to Py3, finally, as of a couple weeks ago.

2. @jcushman that regex engine is interesting. Seems like maybe it'd make a good optional module to swap in once we get the basics going with the current code. How much faster is it?

**Comment by jcushman** [comment]
Hyperscan makes the `tokenize()` function about 30 times as fast on my machine -- [here's a gist](https://gist.github.com/jcushman/d54d3771f54e0d7302fd189f917a9ad2). I don't have a handy way to test what fraction of the time you spend in `tokenize()`, is it a lot?

Hyperscan comes along with a bunch of [weird constraints](https://intel.github.io/hyperscan/dev-reference/compilation.html#semantics), like having a long compile phase, returning every possible overlapping match, and not supporting capture groups, so there's some funky code to deal with all of that, but it works.

I'm still chewing on whether it should change the underlying design that throwing in additional regexes to the tokenizing phase and knowing which ones matched where is essentially free, but it feels like that ought to be helpful? Anyway I agree that for starters it makes sense just to treat this as a future speed upgrade.

**Comment by mlissner** [comment]
Yeah, I don't know how much we use in tokenize either. I guess this will get easier as we start working on pulling things apart. 

In the meantime, I created a [new repo for this project called "microscope."](https://github.com/freelawproject/microscope) I'm not sure I like the name, but something about looking really closely at documents, a pun on finding cites/sites, I dunno. Had to start somewhere, so it exists now. Feels almost like a good enough name.

**Comment by mattdahl** [comment]
Thanks for the replies, everyone. I'm starting to work on this.

What is the status of #1230? I was looking at those court citation strings again (https://github.com/freelawproject/courtlistener/blob/master/cl/citations/find_citations.py#L112) and it seems like the appropriate place to put them could actually be in the standalone `courts_db` package. @mlissner @flooie do you agree? What is the eventual/intended relationship of that package to the data stored in the `Court` CL model?

**Comment by mlissner** [comment]
We haven't done #1230 yet, but we should. I agree that the citation strings would feel very comfy in there. The db itself will continue living on, since it's needed for joins and whatnot. So far, courts_db is mostly about having regexes for court strings.

**Comment by mattdahl** [comment]
Is the set of `id`s in https://github.com/freelawproject/courts-db/blob/master/courts_db/data/courts.json equal/close to equal to the set of `pk`s stored in the CourtListener `Court` DB? I assume that it is, but just checking. If so, I will simply add all the citation strings to their appropriate objects in that `.json. file.

**Comment by mattdahl** [comment]
Actually, I just looked into this myself, and I see 715 courts in `courts_db` but only 458 in CL's fixtures. And 32 of those 458 are apparently not in `courts_db` yet. Do these numbers sound roughly correct? I want to be sure I'm not missing something. I can also post a gist of my merging code or share a list of the courts that are in CL but missing from `courts_db`, if desired.

**Comment by mlissner** [comment]
Yep, that should be reliable, though I'm not sure it's complete. Here's the complete list of IDs: 

    scotus,ca1,ca2,ca3,ca4,ca5,ca6,ca7,ca8,ca9,ca10,ca11,cadc,cafc,ag,afcca,asbca,armfor,acca,cc,uscfc,com,ccpa,cusc,tax,bta,mc,mspb,nmcca,cavc,bva,eca,tecoa,fiscr,fisc,reglrailreorgct,cit,bap1,bap2,bap6,bap8,bap9,bap10,bapme,bapma,almb,alnb,alsb,akb,arb,areb,arwb,cacb,caeb,canb,casb,cob,ctb,deb,dcb,flmb,flnb,flsb,gamb,ganb,gasb,hib,idb,ilcb,ilnb,ilsb,innb,insb,ianb,iasb,ksb,kyeb,kywb,laeb,lamb,lawb,meb,mdb,mab,mieb,miwb,mnb,msnb,mssb,moeb,mowb,mtb,nebraskab,nvb,nhb,njb,nmb,nyeb,nynb,nysb,nywb,nceb,ncmb,ncwb,ndb,ohnb,ohsb,okeb,oknb,okwb,orb,paeb,pamb,pawb,rib,scb,sdb,tneb,tnmb,tnwb,tennesseeb,txeb,txnb,txsb,txwb,utb,vtb,vaeb,vawb,waeb,wawb,wvnb,wvsb,wieb,wiwb,wyb,gub,nmib,prb,vib,dcd,ald,almd,alnd,alsd,akd,azd,ared,arwd,californiad,cacd,caed,cand,casd,caca,cod,ctd,ded,circtdel,fld,flmd,flnd,flsd,gad,gamd,gand,gasd,hid,idd,ilcd,ilnd,ilsd,illinoised,illinoisd,innd,insd,indianad,iad,iand,iasd,ksd,kyd,kyed,kywd,lad,laed,lamd,lawd,orld,med,mdd,mad,michd,mied,miwd,mnd,missd,msnd,mssd,mod,moed,mowd,mtd,ned,nvd,nhd,njd,nmd,nyd,nyed,nynd,nysd,nywd,nced,ncmd,ncwd,circtnc,ndd,ncd,ohnd,ohsd,ohiod,oked,oknd,okwd,ord,paed,pamd,pawd,pennsylvaniad,rid,scd,southcarolinaed,southcarolinawd,sdd,tned,tnmd,tnwd,tennessed,circttenn,texd,txed,txnd,txsd,txwd,utd,vtd,vad,vaed,vawd,washd,waed,wawd,wvad,wvnd,wvsd,wisd,wied,wiwd,wyd,canalzoned,gud,nmid,prd,vid,ala,alactapp,alacrimapp,alacivapp,alaska,alaskactapp,ariz,arizctapp,arizsuper,ariztaxct,ark,arkctapp,arkworkcompcom,arkag,cal,calctapp,calappdeptsuper,calag,colo,coloctapp,coloworkcompcom,coloag,conn,connappct,connsuperct,connworkcompcom,del,delch,delsuperct,delctcompl,delfamct,deljudct,dc,fla,fladistctapp,flaag,ga,gactapp,haw,hawapp,idaho,idahoctapp,ill,illappct,ind,indctapp,indtc,iowa,iowactapp,kan,kanctapp,kanag,ky,kyctapp,kyctapphigh,la,lactapp,laag,me,md,mdctspecapp,mdag,mass,massappct,masssuperct,massdistct,maworkcompcom,mich,michctapp,minn,minnctapp,minnag,miss,missctapp,mo,moctapp,moag,mont,monttc,montag,neb,nebctapp,nebag,nev,nh,nj,njsuperctappdiv,njtaxct,njch,nm,nmctapp,ny,nyappdiv,nyappterm,nysupct,nyfamct,nysurct,nycivct,nycrimct,nyag,nc,ncctapp,ncsuperct,ncworkcompcom,nd,ndctapp,ohio,ohioctapp,ohioctcl,okla,oklacivapp,oklacrimapp,oklajeap,oklacoj,oklaag,or,orctapp,ortc,pa,pasuperct,pacommwct,cjdpa,ri,risuperct,sc,scctapp,sd,tenn,tennctapp,tenncrimapp,tennworkcompcl,tennworkcompapp,tennsuperct,tex,texapp,texcrimapp,texreview,texjpml,texag,utah,utahctapp,vt,vtsuperct,va,vactapp,wash,washctapp,washag,wva,wis,wisctapp,wisag,wyo,usjc,jpml,sttex,stp,kingsbench,test,psc

When you add the additional field to the json file, can you also document it in the readme?

**Comment by mlissner** [comment]
My comment was from before I saw yours. The list in my comment should be complete. I could see some courts being missing from either source. The courts DB has a bunch that we're going to have to add to CL fairly soon, as we do a huge import.

**Comment by mlissner** [comment]
I think it's safe to say we put the citation parser in its own package: freelawproject/eyecite lives. Thank you everybody.
